client/dashboard/domains/domain-forwarding/utils.ts

```diff
@@ -0,0 +1,134 @@
+// Validation helpers for domain forwarding form
+
+import type { FormData } from './form';
+import type { DomainForwarding, DomainForwardingSaveData } from '../../../data/domain-forwarding';
+
+const SOFT_URL_REGEX =
+	/^(https?:)?(?:[a-zA-Z0-9-.]+\.)?[a-zA-Z]{0,}(?:\/[^\s]*){0,}(:[0-9/a-z-#]*)?$/i;
+const SUBDOMAIN_LABEL_REGEX = /^(?!-)[a-zA-Z0-9-]{0,62}[a-zA-Z0-9]$|^[a-zA-Z0-9]$/i;
+
+export function normalizeUrlWithDefaultProtocol( input: string ): string {
+	if ( ! input ) {
+		return '';
+	}
+	return /^https?:\/\//i.test( input ) ? input : `https://${ input }`;
+}
+
+function getDomainPartsAndLevel( domain: string ): [ string[], number ] {
+	const parts = domain.split( '.' );
+	return [ parts, parts.length ];
+}
+
+function isSameDomain(
+	domainParts: string[],
+	targetHostParts: string[],
+	domainLevel: number
+): boolean {
+	for ( let i = 0; i < domainLevel; i++ ) {
+		if ( domainParts[ i ] !== targetHostParts[ i ] ) {
+			return false;
+		}
+	}
+	return true;
+}
+
+function isSubdomain(
+	domainParts: string[],
+	targetHostParts: string[],
+	domainLevel: number,
+	targetHostLevel: number
```

Comment:
Is it possible to take `domainParts` and `targetHostParts` (or `sourceDomain` and `targetDomain`) as argument? Looks like we can get the level from them. Is it correct?